### PR TITLE
Various cleanups

### PR DESCRIPTION
**Pull request by fzaiser**
By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).

### CODE REVIEW

File: src/ProbabilisticProgramming/Independence.dfy [hunk 1]
```diff
@@ -52,6 +52,16 @@ module Independence {
   lemma {:axiom} ReturnIsIndep<T>(x: T)
     ensures IsIndep(Monad.Return(x))
 
+  lemma MapIsIndep<A, B(!new)>(f: Monad.Hurd<A>, g: A -> B)
```
Comment (stefan-aws) [inline src/ProbabilisticProgramming/Independence.dfy:55]: Nice, thank you.

File: src/ProbabilisticProgramming/Loops.dfy [hunk 2]
```diff
@@ -40,20 +40,20 @@ module Loops {
   // This definition is opaque because the details are not very useful.
   // For proofs, use the lemma `WhileUnroll`.
   // Equation (3.25), but modified to use `Monad.Diverging` instead of HOL's `arb` in case of nontermination
-  // TODO: While(condition, body)(init) would be cleaner
-  opaque ghost function While<A>(condition: A -> bool, body: A -> Monad.Hurd<A>, init: A): (f: Monad.Hurd<A>)
-    ensures forall s: Rand.Bitstream :: !condition(init) ==> f(s) == Monad.Return(init)(s)
+  opaque ghost function While<A(!new)>(condition: A -> bool, body: A -> Monad.Hurd<A>): (loop: A -> Monad.Hurd<A>)
```
Comment (stefan-aws) [inline src/ProbabilisticProgramming/Loops.dfy:43]: I can't quite see what has caused this change yet. How does it make our life easier if we use this curried version of the current `While`?

File: src/Distributions/DiscreteLaplace/Implementation.dfy [hunk 3]
```diff
@@ -37,7 +37,7 @@ module DiscreteLaplace.Implementation {
         }
         var x := u + scale.numer * v;
         y := x / scale.denom;
-        b := BernoulliSample(Rationals.Rational(1, 2));
+        b := CoinSample();
```
Comment (stefan-aws) [inline src/Distributions/DiscreteLaplace/Implementation.dfy:40]: Good catch.

File: src/ProbabilisticProgramming/Loops.dfy (hunk 2 above)
Comment (fzaiser) [reply src/ProbabilisticProgramming/Loops.dfy:43]: I think it's easier to view a while loop as transforming a value `a: A`. If you look at the diff, it was used as a function `(init: A) => While(condition, body, init)` in many places in the file. The curried version can just be written as `While(condition, body)`.

Comment (fzaiser) [reply src/ProbabilisticProgramming/Loops.dfy:43]: But I'm not super attached to that change, so if you don't like it, I can undo it.

Comment (stefan-aws) [reply src/ProbabilisticProgramming/Loops.dfy:43]: Ah, makes sense!

**Review by stefan-aws — APPROVED** [review]

**Pull request merged**